### PR TITLE
guilib: add static helpers to remove ifdeffing

### PR DESCRIPTION
**Pull request by lrusak**
I've been working on this for a while and is basically just a split off of #15549 

This adds static create functions for DX and GL classes. As it's likely these two types will never be built together we can do this without a factory.

There are some problems with polymophism in that we need to be able to copy a class from the base class. That's where the `Clone()` method comes in. This seems like the best way to do this.

This also uses `unique_ptr` which makes it so we have to define copy constructors for any class that uses it for a member. I have declared the copy constructors protected or private where I can to disable their use otherwise.

I've tried to split the commits as nicely as possible. functional changes are separate from renames (the renames aren't really necessary but I think it helps).

This is also something that I'm fine not having merged in if people are against it. If the changes rot and rebases are needed it's likely that things will break as some copy constructors are quite extensive and changes may go unnoticed.

Basically it comes down to removing some ugly ifdeffing and fixes some inheritance problems. Is it worth it? I don't know.

Let me know what you think.

### CODE REVIEW

**Review by garbear — APPROVED** [review]
Clone() is a useful idiom for copy-constructing polymorphic classes. Good job hiding the copy constructors.

I appreciated the rename commits separate from the rest. On my first review pass, I actually start with reading the files modified list. You can tell a lot about what dependencies you're dealing with - files modified outside of guilib evidently have a dependency on guilib, and it's a step to evaluate if this dependency is proper. In all cases I see, it is.

Furthermore, reading the diff of the renames, I can see the uses of the subject across the entire codebase. This factors further into my decision of whether the dependency is proper, and if the external class is using the API correctly.

When I built RetroPlayer, the crazy ifdeffing you fix here was an impediment. I'm glad for your work and think it should be merged.

**Review by yol — APPROVED** [review]
Nice! Clone() is a perfectly valid solution IMO and if I remember correctly a standard pattern.

Would it make sense to go even further and return `std::unique_ptr` from Clone() and Create*()?

**Comment by lrusak** [comment]
> Clone() is a useful idiom for copy-constructing polymorphic classes. Good job hiding the copy constructors.
> 
> I appreciated the rename commits separate from the rest. On my first review pass, I actually start with reading the files modified list. You can tell a lot about what dependencies you're dealing with - files modified outside of guilib evidently have a dependency on guilib, and it's a step to evaluate if this dependency is proper. In all cases I see, it is.
> 
> Furthermore, reading the diff of the renames, I can see the uses of the subject across the entire codebase. This factors further into my decision of whether the dependency is proper, and if the external class is using the API correctly.
> 
> When I built RetroPlayer, the crazy ifdeffing you fix here was an impediment. I'm glad for your work and think it should be merged.

Thanks for the kind words :+1: I always love your reviews.


> Nice! Clone() is a perfectly valid solution IMO and if I remember correctly a standard pattern.
> 
> Would it make sense to go even further and return `std::unique_ptr` from Clone() and Create*()?

Yes I think it would, however, I think the changes would be better suited for a follow up PR as the changes here are quite extensive already :smile:

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
objections?
